### PR TITLE
[Ansor][AutoTVM v2.0] Phase 1: The base class for cost models

### PR DESCRIPTION
**Pull request by merrymercy**
For the full upstream plan, see [Ansor RFC](https://discuss.tvm.ai/t/rfc-ansor-an-auto-scheduler-for-tvm-autotvm-v2-0/7005/32).

This PR adds the base class for cost models and a random cost model.
Because learning-based models typically use python packages (e.g., xgboost, pytorch, tensorflow), we prefer to implement them in python and use the c++ wrapper PythonBasedCostModel for calling them in c++.

The later PR will introduce a cost model based on manual features and xgboost.

### CODE REVIEW

**Comment by merrymercy** [comment]
cc @junrushao1994 @comaniac @tqchen @jcf94

**Comment by junrushao1994** [comment]
Will review Tuesday night

File: include/tvm/auto_scheduler/cost_model.h [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+                       std::vector<float>* scores) = 0;
+
+  /*!
+   * \brief Predict the scores of all stages in states
+   * \param task The search task
+   * \param states The input states
+   * \param state_scores The predicted scores for all states
+   * \param stage_scores The predicted scores for all stages in all stages
+   */
+  virtual void PredictStages(const SearchTask& task, const std::vector<State>& states,
```
Comment (jcf94) [inline include/tvm/auto_scheduler/cost_model.h:67]: Do we still need this part? Since out current search policy only apply cost model on a complete schedule.

Comment (comaniac) [reply include/tvm/auto_scheduler/cost_model.h:67]: This is needed by crossover, and it should be useful for future developments IMO.

File: python/tvm/auto_scheduler/cost_model/cost_model.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        scores: List[float]
+            The predicted scores for all states
+        """
+        return [x.value for x in _ffi_api.CostModelPredict(self, search_task, states)]
+
+
+@tvm._ffi.register_func("auto_scheduler.cost_model.random_number")
+def random_number(n, return_ptr):
+    """ A random number generator func for c++'s RandomModel """
+    if n == 0:
```
Comment (comaniac) [inline python/tvm/auto_scheduler/cost_model/cost_model.py:70]: Better to improve the naming of `n`.

File: python/tvm/auto_scheduler/cost_model/cost_model.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+        Returns
+        -------
+        scores: List[float]
+            The predicted scores for all states
+        """
+        raise NotImplementedError
+
+    def predict_stages(self, task, states):
+        """Predict the scores of states
```
Comment (comaniac) [inline python/tvm/auto_scheduler/cost_model/cost_model.py:128]:
```suggestion
        """Predict the scores of states. Different from "predict", this function returns score breakdown for each stage.
```

File: include/tvm/auto_scheduler/cost_model.h (hunk 1 above)
Comment (merrymercy) [reply include/tvm/auto_scheduler/cost_model.h:67]: Yeah, your suggestion seems to be better.

File: tests/python/unittest/test_auto_scheduler_cost_model.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+"""Test cost models"""
+
+import tvm
+from tvm import auto_scheduler
+
+from test_auto_scheduler_common import matmul_auto_scheduler_test
+
+
+def test_random_model():
```
Comment (FrozenGene) [inline tests/python/unittest/test_auto_scheduler_cost_model.py:26]:
```suggestion
def test_random_model():
    if not tvm.runtime.enabled("llvm"):
        return
```

Comment (FrozenGene) [reply tests/python/unittest/test_auto_scheduler_cost_model.py:26]: Final comment. Let us disable this test if users' environment doesn't have llvm.
